essentials/src/api/api_client.rs

```diff
@@ -15,66 +15,249 @@
+			let addr = polkadot::storage().hrmp().hrmp_channels(&id);
+			self.storage()
+				.at(block_hash)
+				.fetch(&addr)
+				.await?
+				.map(|hrmp_channel_configuration| {
+					channels_configuration.insert(peer_parachain_id, hrmp_channel_configuration.into())
+				});
+		}
+		Ok(channels_configuration)
+	}
+
+	async fn fetch_dynamic_storage(
+		&self,
+		maybe_hash: Option<H256>,
+		pallet_name: &str,
+		entry_name: &str,
+	) -> Result<Option<Value<u32>>, subxt::Error> {
+		let storage = match maybe_hash {
+			Some(hash) => self.storage().at(hash),
+			None => self.storage().at_latest().await?,
+		};
+		match storage
+			.fetch(&subxt::dynamic::storage(pallet_name, entry_name, Vec::<u8>::new()))
+			.await?
+		{
+			Some(v) => Ok(Some(v.to_value()?)),
+			None => Ok(None),
+		}
+	}
+
+	async fn extract_parainherent(&self, maybe_hash: Option<H256>) -> Result<InherentData, subxt::Error> {
+		let block = self.block_at(maybe_hash).await?;
+		let ex = block
+			.extrinsics()
+			.await?
+			.iter()
+			.take(2)
+			.last()
+			.expect("`ParaInherent` data is always at index #1")
+			.expect("`ParaInherent` data must exist");
+		let enter = ex
+			.as_extrinsic::<polkadot::para_inherent::calls::types::Enter>()
+			.expect("Failed to decode `ParaInherent`")
+			.expect("`ParaInherent` must exist");
+
+		Ok(enter.data)
+	}
+
+	async fn legacy_get_block_hash(
 		&self,
 		maybe_block_number: Option<BlockNumber>,
 	) -> Result<Option<H256>, subxt::Error> {
 		let maybe_block_number = maybe_block_number.map(|v| NumberOrHex::Number(v.into()));
 		self.legacy_rpc_methods.chain_get_block_hash(maybe_block_number).await
 	}
 
-	pub async fn stream_best_block_headers(&self) -> Result<HeaderStream, subxt::Error> {
+	async fn stream_best_block_headers(&self) -> Result<HeaderStream, subxt::Error> {
 		self.client.backend().stream_best_block_headers().await
 	}
 
-	pub async fn stream_finalized_block_headers(&self) -> Result<HeaderStream, subxt::Error> {
+	async fn stream_finalized_block_headers(&self) -> Result<HeaderStream, subxt::Error> {
 		self.client.backend().stream_finalized_block_headers().await
 	}
 }
+
+pub async fn build_online_client(url: &str) -> Result<ApiClient<OnlineClient<PolkadotConfig>>, String> {
+	let rpc_client = RpcClient::from_url(url)
+		.await
+		.map_err(|e| format!("Cannot construct RPC client: {e}"))?;
+	let client = OnlineClient::from_rpc_client(rpc_client.clone())
+		.await
+		.map_err(|e| format!("Cannot construct OnlineClient from rpc client: {e}"))?;
+	let legacy_rpc_methods = LegacyRpcMethods::<PolkadotConfig>::new(rpc_client);
+
+	Ok(ApiClient { client, legacy_rpc_methods })
+}
+
+pub async fn build_light_client(url: &str) -> Result<ApiClient<LightClient<PolkadotConfig>>, String> {
+	let rpc_client = RpcClient::from_url(url)
+		.await
+		.map_err(|e| format!("Cannot construct RPC client: {e}"))?;
+	let client = LightClient::builder()
+		.build_from_url(url)
```

Comment:
We still need an RPC endpoint here, why not use `build` from chainspec ?

Reply:
We can, but why it's better? We need only one call to download the chainspec. It's easy to connect to other networks. Otherwise, we need to have prepared a chainspecs for every particular network.

Reply:
The point was to not need an RPC node at all. Once the light client works fully, we should only use RPC to get a chainspec

Reply:
Unfortunately, chainspec changes over the time: fields `babeFinalizedBlockWeight`. `finalizedBlockHeader`, `grandpaAuthoritySet`. So it's better to fetch the fresh one when we connect.

Reply:
I see,  let’s download it for now